### PR TITLE
Restful API for orders

### LINKED ISSUE #101: Refactor: refresh views should return proper JSON, not HTML-in-JSON
Even though it is a nice feature, it is bad coding style in my opinion. The JSON API should deliver proper JSON

### LINKED ISSUE #125: Enhancement: More structured POST views
The POST views that handle changing shift parameters via javascript calls (AJAX) are now mixed up and not intuitive. We need to refactor all views that handle AJAX calls and make them function the same.

This might also directly close #101

### PR DESCRIPTION
**Pull request by KiOui**
Added a Restful API for orders application.

Added the following:

- API views for listing, creating, retrieving and updating Shift objects
- API views for listing, creating, retrieving, updating and destroying Order objects
- API views for listing Product objects
- API views for the features available on the Shift admin screen
- Serializers for all the above methods
- VueJS to easily manage the state of the HTML document in the browser
- A CustomAutoSchema to make adjustments to the provided AutoSchema
- SwaggerUI

Removed the following:

- All views that we were using to provide API-like features that are now implemented in the API
- A couple of templates that are now included in bigger ones to optimize the VueJS and API implementation

To test you can visit `/api/docs`.

Closes #125 
Closes #101

### CODE REVIEW

File: website/orders/api/serializers/orders.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        :param validated_data: validated serialized data
+        :return: an Order if the Order could be successfully added, a ValidationError otherwise
+        """
+        try:
+            return add_order(
+                validated_data["product"],
+                validated_data["shift"],
+                validated_data["type"],
+                user=validated_data["user"],
+                paid=validated_data["paid"],
```
Comment (JobDoesburg) [inline website/orders/api/serializers/orders.py:45]: shouldn't the API always create unpaid and unready orders, which an admin can then make paid and ready with a separate PATCH?

File: website/orders/api/serializers/orders.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            "paid_at",
+            "type",
+        ]
+        read_only_fields = ["pk", "created", "user", "order_price", "ready_at", "paid_at"]
+
+
+class ShiftSerializer(serializers.ModelSerializer):
+    """Serializer for Shift objects."""
+
+    def validate(self, data):
```
Comment (JobDoesburg) [inline website/orders/api/serializers/orders.py:76]: is this logic required here? can't we use the clean-method from the model?

File: website/tosti/settings/settings.py [hunk 3]
```diff
@@ -38,6 +38,7 @@
     "admin_auto_filters",
     "import_export",
     "guardian",
+    'rest_framework',
```
Comment (JobDoesburg) [inline website/tosti/settings/settings.py:41]: Can we add swaggerui?

**Review by JobDoesburg — CHANGES_REQUESTED** [review]
How does authentication work?

File: website/tosti/settings/settings.py (hunk 3 above)
Comment (se-bastiaan) [reply website/tosti/settings/settings.py:41]: This sounds like a marvelous idea

Comment (JobDoesburg) [reply website/tosti/settings/settings.py:41]: @KiOui bump

File: website/orders/api/v1/views.py [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+        API endpoint for creating a Shift.
+        Permission required: can_manage_shift_in_venue
+        :param request: the request
+        :param args: arguments
+        :param kwargs: keyword arguments
+        :return: super method or PermissionDenied when 'venue' POST data is None or when user does not have permission
+        in indicated venue
+        """
+        venue = request.data.get("venue")
+        if request.user.has_perm("orders.can_manage_shift_in_venue", venue):
```
Comment (JobDoesburg) [inline website/orders/api/v1/views.py:214]: don't we have a nice permission class for this?

File: website/orders/api/v1/views.py [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+
+    serializer_class = ProductSerializer
+    queryset = Product.objects.all()
+
+    def get_queryset(self):
+        """Get the queryset."""
+        return self.queryset.filter(available_at=self.kwargs.get("shift").venue)
+
+
+@api_view(["POST"])
```
Comment (JobDoesburg) [inline website/orders/api/v1/views.py:256]: I think it would be nicer to make this a PUT

File: website/orders/services.py [hunk 6]
```diff
@@ -97,8 +99,8 @@ def add_order(product, shift, order_type, user=None, set_done=False, force=False
             type=order_type,
             user=user,
             user_association=user.profile.association,
-            paid=set_done,
-            ready=set_done,
+            paid=paid,
```
Comment (JobDoesburg) [inline website/orders/services.py:102]: Again, do we need to set 'paid' and 'ready'?

File: website/orders/api/v1/urls.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+urlpatterns = [
+    path("shifts", ShiftListCreateAPIView.as_view(), name="shifts_listcreate"),
+    path("shifts/<int:pk>", ShiftRetrieveUpdateAPIView.as_view(), name="shift_retrieveupdate"),
+    path("shifts/<shift:shift>/orders", OrderListAPIView.as_view(), name="orders_list"),
+    path(
+        "shifts/<shift:shift>/orders/<int:pk>",
+        OrderRetrieveUpdateDestroyAPIView.as_view(),
+        name="orders_retrieveupdatedestroy",
+    ),
```
Comment (JobDoesburg) [inline website/orders/api/v1/urls.py:28]: I think it would be nice to add an endpoint /shift/<shift:shift>/orders/<int:pk>/ready to which you can PUT to make the order ready, and idem with paid.

File: website/orders/api/v1/urls.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    path("shifts/<shift:shift>/orders/add", OrderCreateAPIView.as_view(), name="orders_create"),
+    path("shifts/<shift:shift>/add-time", shift_add_time, name="shifts_add_time"),
+    path("shifts/<shift:shift>/add-capacity", shift_add_capacity, name="shifts_add_capacity"),
+    path("shifts/<shift:shift>/scanner", shift_scanner, name="shifts_scanner"),
+    path("shifts/<shift:shift>/cart-order", OrderView.as_view(), name="shifts_cart_order"),
```
Comment (JobDoesburg) [inline website/orders/api/v1/urls.py:33]: Do we really need this separately? Can't we just leave the whole cart concept to the user/browser and allow batch requests to /orders/add?

**Review by JobDoesburg — CHANGES_REQUESTED** [review]

File: website/orders/api/v1/urls.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+urlpatterns = [
+    path("shifts", ShiftListCreateAPIView.as_view(), name="shifts_listcreate"),
+    path("shifts/<int:pk>", ShiftRetrieveUpdateAPIView.as_view(), name="shift_retrieveupdate"),
+    path("shifts/<shift:shift>/orders", OrderListAPIView.as_view(), name="orders_list"),
+    path(
+        "shifts/<shift:shift>/orders/<int:pk>",
+        OrderRetrieveUpdateDestroyAPIView.as_view(),
+        name="orders_retrieveupdatedestroy",
+    ),
+    path("shifts/<shift:shift>/orders/add", OrderCreateAPIView.as_view(), name="orders_create"),
```
Comment (se-bastiaan) [inline website/orders/api/v1/urls.py:29]: If you put the OrderCreateAPIView and the OrderListAPIView together you can properly implement the creation of orders as a POST request. Decide what you do based on the HTTP method and not the route.

File: website/orders/api/v1/urls.py (hunk 7 above)
Comment (KiOui) [reply website/orders/api/v1/urls.py:28]: This can be done with a PATCH to the endpoint for a specific Order object.

File: website/orders/api/v1/urls.py (hunk 8 above)
Comment (KiOui) [reply website/orders/api/v1/urls.py:33]: We can do that but that makes the Javascript ordering a bit more complicated, and this works as it is right now, so let's just use what we have right now.

File: website/orders/api/v1/urls.py (hunk 9 above)
Comment (KiOui) [reply website/orders/api/v1/urls.py:29]: I did it this way because the OrderRetrieveUpdateDestroy view and the OrderCreate view use different serializers because of the way we create Order objects in the backend. It seemed more natural to create a different API endpoint for both views as they do not rely on the same set of data that they let the API user access or create.

File: website/orders/api/v1/views.py (hunk 4 above)
Comment (KiOui) [reply website/orders/api/v1/views.py:214]: Don't think so.

File: website/orders/api/v1/serializers.py [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+            "ready_at",
+            "paid",
+            "paid_at",
+            "type",
+        ]
+        read_only_fields = ["id", "created", "user", "product", "order_price", "ready_at", "paid_at", "type"]
+
+
+class UserSerializer(serializers.RelatedField):
+    """User serializer."""
```
[the remark this answers is not in the document]
Comment (KiOui) [reply website/orders/api/v1/serializers.py]: Done

File: website/orders/api/v1/serializers.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+        model = User
+        fields = [
+            "id",
+            "username",
+        ]
+        read_only_fields = ["id", "username"]
+
+
+class OrderCreateSerializer(serializers.ModelSerializer):
+    """Serializer for creation of Order."""
```
[the remark this answers is not in the document]
Comment (KiOui) [reply website/orders/api/v1/serializers.py:51]: The OrderCreateAPIView and the OrderRetrieveUpdateDestroyAPIView rely on different pieces of data. For example, the create view only needs a user, product, and type of Order whereas the retrieve of an Order shows the "product" as an "icon" and a "name" attribute. When creating an Order the "product" must be supplied as an "id" only. This difference seemed a bit off and I thus decided to split them up.

File: website/orders/api/v1/serializers.py [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+
+        model = models.Product
+        fields = [
+            "name",
+            "icon",
+        ]
+        read_only_fields = ["name", "icon"]
+
+
+class UserRelatedField(serializers.ModelSerializer):
```
[the remark this answers is not in the document]
Comment (KiOui) [reply website/orders/api/v1/serializers.py]: Moved to the users application

File: website/orders/api/v1/permissions.py [hunk 13]
```diff
@@ -0,0 +1,9 @@
+from rest_framework.permissions import BasePermission
+
+
+class HasGuardianPermission(BasePermission):
```
[the remark this answers is not in the document]
Comment (KiOui) [reply website/orders/api/v1/permissions.py]: Refactored

File: website/orders/services.py (hunk 6 above)
Comment (KiOui) [reply website/orders/services.py:102]: I reverted it to setting paid and ready later.

File: website/orders/api/v1/views.py (hunk 4 above)
Comment (KiOui) [reply website/orders/api/v1/views.py:214]: Nope

File: website/orders/api/v1/views.py (hunk 5 above)
Comment (KiOui) [reply website/orders/api/v1/views.py]: I altered it to PATCH

File: website/tosti/settings/settings.py (hunk 3 above)
Comment (KiOui) [reply website/tosti/settings/settings.py:41]: Done

**Comment by JobDoesburg** [comment]
- [ ] We still need to style swagger UI with some TOSTI styling
- [ ] In my opinion the API should accept POST to `/orders` instead of POST to `/cart-order/` or `/orders/add`
- [ ] We should add / improve documentation to the schema

**Comment by KiOui** [comment]
> * [ ]  In my opinion the API should accept POST to `/orders` instead of POST to `/cart-order` or `/orders/add`

The `/cart-order` endpoint will still be necessary, I'll try to add the POST to the `/orders` endpoint.

**Comment by KiOui** [comment]
We also need to implement token support for the authenticator

File: website/orders/api/v1/views.py [hunk 14]
```diff
@@ -0,0 +1,393 @@
...
+    def post(self, request, **kwargs):
+        """
+        Create multiple Orders in one go.
+
+        API endpoint for creating multiple orders in one go (handling of a cart).
+        A "cart" POST parameter must be specified including the ID's of the Products the user wants to order.
+        Permission required: can_order_in_venue
+        :param request: the request
+        :param kwargs: keyword arguments
+        :return: a 201 response code if all orders in the cart were successfully processed. Raises a ValidationError
```
Comment (JobDoesburg) [inline website/orders/api/v1/views.py:69]: We have pydoc `:raises:`

File: website/orders/services.py [hunk 15]
```diff
@@ -86,7 +88,7 @@ def add_order(product, shift, order_type, user=None, set_done=False, force=False
             pass
         # Shift checks
         if not shift.can_order:
-            raise OrderException("User can not order products for this shift")
+            raise OrderException("This shift is closed.")
         if not shift.is_active:
             raise OrderException("This shift is not active")
```
Comment (JobDoesburg) [inline website/orders/services.py:93]: Isn't this checked by the model?

File: website/tosti/api/openapi.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
+from oauth2_provider.scopes import get_scopes_backend
+from rest_framework.reverse import reverse
+from rest_framework.schemas.openapi import SchemaGenerator, AutoSchema
+
+
+class OAuthSchemaGenerator(SchemaGenerator):
```
Comment (JobDoesburg) [inline website/tosti/api/openapi.py:6]: Shouldn't this be called the OpenAPISchemaGenerator instead of OAuth (as it is openapi schema)

File: website/marietje/api/v1/views.py [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+def player_play(request, **kwargs):
+    """
+    Player Play API View.
+
+    Permission required: marietje.can_request
+
+    Start playback on a Player.
+    """
+    player = kwargs.get("player")
+    if request.user.has_perm("marietje.can_request", player):
```
Comment (JobDoesburg) [inline website/marietje/api/v1/views.py:123]: shouldn't this be 'can manage'?

File: website/marietje/api/v1/views.py [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+    serializer_class = QueueItemSerializer
+    queryset = SpotifyQueueItem.objects.all()
+    pagination_class = StandardResultsSetPagination
+
+    def get_queryset(self):
+        """Get the queryset."""
+        return self.queryset.filter(player=self.kwargs.get("player"))
+
+
+@api_view(["GET"])
```
Comment (JobDoesburg) [inline website/marietje/api/v1/views.py:59]: If we could rewrite these to class-based views, you could add the parameters like the search query to the openapi schema so swagger would understand it (by setting `schema=` for the view)

**Comment by KiOui** [comment]
As stated [here](https://github.com/encode/django-rest-framework/commit/7915485c0d94c1ca9ebe6606b78465fffe45acac), there is no easy way to implement manual schema adjustments to AutoSchema of OpenAPI for now, so we have to write our own.

File: website/marietje/api/v1/views.py (hunk 17 above)
Comment (KiOui) [reply website/marietje/api/v1/views.py:129]: can_control actually but nice catch!

File: website/marietje/api/v1/views.py (hunk 18 above)
Comment (KiOui) [reply website/marietje/api/v1/views.py]: Done, only the Player schema is not displayed correctly but we might fix that later.

File: website/orders/api/v1/views.py (hunk 14 above)
Comment (KiOui) [reply website/orders/api/v1/views.py]: I have removed most of them as they were displayed on the SwaggerUI page (and that was ugly).

File: website/orders/api/v1/serializers.py [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+
+
+class OrderSerializer(serializers.ModelSerializer):
+    """Serializer for Order model."""
+
+    product_id = serializers.PrimaryKeyRelatedField(queryset=Product.objects.all(), many=False, read_only=False)
+    product = ProductSerializer(many=False, read_only=True)
+    user = UserRelatedField(many=False, read_only=True)
+
+    def validate_product_id(self, value):
```
[the remark this answers is not in the document]
Comment (KiOui) [reply website/orders/api/v1/serializers.py:50]: What do you mean?

File: website/orders/services.py (hunk 15 above)
Comment (KiOui) [reply website/orders/services.py:93]: We have to rewrite this thing anyway, as it does not work completely (see the TODO somewhere above this line) so let's do this in a different pull request.

File: website/marietje/api/v1/views.py [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+        Use this endpoint to search for a Spotify Track. Tracks can be searched via their Spotify id.
+        """
+        player = kwargs.get("player")
+        query = request.GET.get("query", "")
+        request_id = request.GET.get("id", None)
+        try:
+            maximum = int(request.GET.get("maximum", 5))
+        except ValueError:
+            maximum = 5
+        if request.user.has_perm("marietje.can_request", player):
```
Comment (JobDoesburg) [inline website/marietje/api/v1/views.py:101]: APIView has a dedicated `check_permissions` / `check_object_permissions` method for this which I think we should use (see http://www.cdrf.co/3.9/rest_framework.views/APIView.html)

File: website/marietje/api/v1/views.py [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+                    return Response(status=status.HTTP_503_SERVICE_UNAVAILABLE)
+                return Response(status=status.HTTP_200_OK)
+            else:
+                raise ValidationError("A track id is required.")
+        else:
+            raise PermissionDenied
+
+
+@api_view(["PATCH"])
+def player_play(request, **kwargs):
```
Comment (JobDoesburg) [inline website/marietje/api/v1/views.py:142]: For consistency I think we should also turn this into class based apiviews

**Review by JobDoesburg — CHANGES_REQUESTED** [review]

**Review by JobDoesburg — APPROVED** [review]
Ja opzicht nice, we moeten alleen ff een issue openen om wat uitgebreidere documentatie hier toe te voegen en OAuth scopes toe te voegen.
